---
title: '190110'
author: "zhuhz"
date: "2019/1/10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## map

```{r}
library(tidyverse)
library(reshape2)
library(maptools)
library(plyr)

china1 <- readShapePoly("CHN_adm/CHN_adm1.shp", proj4string=CRS("+proj=longlat +ellps=clrk66"))

num.prov <- length(china1$NAME_1)

x1 <- china1@data
xs1 <- data.frame(x1,id = seq(0:31)-1)

china_map1 <- fortify(china1)

china_mapdata1 <- join(china_map1, xs1, type = "full")

subset <- c("Guizhou","Chongqing","Sichuan","Yunnan","Guangxi")

china_mapdata1 <- china_mapdata1 %>%
  mutate(subset = ifelse(NAME_1%in%subset,"yes","no"))

samp_area.p <- china_mapdata1 %>%
  ggplot(aes(x = long,y = lat,group = group,fill = subset)) + theme_void() + geom_polygon() + geom_path(color = "gray40") + scale_fill_manual(values = c("white","gray"),guide = FALSE)
samp_area.p
ggsave(samp_area.p,filename = "samp_area.pdf",height = 4,width = 5)

samp_prov <- china_mapdata1 %>%
  filter(NAME_1%in%subset)

samp_prov.all <- samp_prov %>%
  ggplot(aes(x = long,y = lat,group = group)) + theme_bw() + geom_polygon(fill = "white") + geom_path(color = "gray40") + annotate("point",x = c(108.0006,108.006,110.5105,108.9311,105.1393,106.8783,103.3825,102.8481),y = c(29.58503,29.59083,24.94182,23.41,28.19017,30.41,25.13444,24.46986),color = "red",size = 2) + labs(x = "Longitude",y = "Latitude") + xlim(97,116)
samp_prov.all
ggsave(samp_prov.all,filename = "samp_prov_all.pdf",height = 4,width = 6)
```

# 16S rRNA gene amplicon analysis

## remove mitochondria and chloroplast

```{r}
counts.in <- read.table("cave.in_otu_table_filtered_sorted.txt", header = T, row.names = 1, sep = "\t" )

counts_clean.in <- counts.in %>%
  mutate(otuID = row.names(.)) %>%
  mutate(keep = ifelse(str_detect(counts.in$taxonomy,"mitochondria")|str_detect(counts.in$taxonomy,"Chloroplast"), "no", "yes")) %>%
  filter(keep == "yes") %>%
  column_to_rownames("otuID") %>%
  select(-keep)

summarise(counts.in,n()) # 11538

summarise(counts_clean.in,n()) # 11444

counts_clean.in %>%
  write.table("cave.in_otu_table_filtered_sorted_clean.txt", sep = "\t")
```

## geography

```{r}
library(FSA)

env <- read.table("environment factors.txt",header = T,sep = "\t") %>%
  group_by(Cave) %>%
  mutate(mean.T = mean(Temperature),sd.T = sd(Temperature),mean.H = mean(Humidity),sd.H = sd(Humidity),mean.pH = mean(pH),sd.pH = sd(pH))

shapiro.test(env$Temperature) #p-value = 3.98e-12
shapiro.test(env$Humidity) #p-value = 4.253e-11
shapiro.test(env$pH) #p-value = 0.02702

kruskal.test(Temperature ~ Cave, env)
#chi-squared = 114.13, df = 7, p-value < 2.2e-16
dunnTest(Temperature ~ Cave,env,method = "none")
#Comparison       P.unadj
#    C1 - C2  0.3279276
#    C1 - G1  1.373681e-11
#    C2 - G1  9.766875e-14
#    C1 - G3  7.915850e-11
#    C2 - G3  7.097480e-13
#    G1 - G3  1.000000
#    C1 - S7  0.5448354
#    C2 - S7  0.1302093
#    G1 - S7  4.260534e-09
#    G3 - S7  1.437545e-08
#    C1 - S8  1.421863e-02
#    C2 - S8  1.131271e-03
#    G1 - S8  1.340768e-04
#    G3 - S8  2.176622e-04
#    S7 - S8  0.072416882
#    C1 - Y2  3.672247e-05
#    C2 - Y2  8.664643e-07
#    G1 - Y2  7.559405e-03
#    G3 - Y2  1.000890e-02
#    S7 - Y2  8.508490e-04
#    S8 - Y2  0.1781873
#    C1 - Y3  3.519648e-05
#    C2 - Y3  1.387709e-06
#    G1 - Y3  0.1047205
#    G3 - Y3  0.1142048
#    S7 - Y3  4.782695e-04
#    S8 - Y3  0.07251945
#    Y2 - Y3  0.5036242

kruskal.test(Humidity ~ Cave, env)
#chi-squared = 80.933, df = 7, p-value = 8.886e-15
dunnTest(Humidity ~ Cave,env,method = "none")
#  Comparison    P.unadj
#    C1 - C2   0.3490266
#    C1 - G1   6.053070e-04
#    C2 - G1   1.483091e-02
#    C1 - G3   1.333230e-10
#    C2 - G3   7.648324e-08
#    G1 - G3   3.214179e-03
#    C1 - S7   0.4506160
#    C2 - S7   0.8715552
#    G1 - S7   1.045423e-02
#    G3 - S7   4.791306e-08
#    C1 - S8   0.9029859
#    C2 - S8   0.4677707
#    G1 - S8   2.748219e-03
#    G3 - S8   8.798359e-09
#    S7 - S8   0.5717796
#    C1 - Y2   3.282786e-06
#    C2 - Y2   3.427545e-04
#    G1 - Y2   0.2996201
#    G3 - Y2   4.277988e-02
#    S7 - Y2   2.235499e-04
#    S8 - Y2   4.591709e-05
#    C1 - Y3   1.310456e-06
#    C2 - Y3   7.800616e-05
#    G1 - Y3   0.07371421
#    G3 - Y3   0.3880533
#    S7 - Y3   5.192268e-05
#    S8 - Y3   1.143721e-05
#    Y2 - Y3   0.3553551

kruskal.test(pH ~ Cave, env)
#chi-squared = 4.3577, df = 7, p-value = 0.7378
```

## profiling

```{r}
map.in <- read.table("mapping.txt",header = T, sep = "\t")

phy <- read.table("in.taxa_summary/cave.in_otu_table_filtered_even_sorted_L2.txt",header = T,row.names = 1,sep = "\t") %>%
  rownames_to_column("taxa")

phy.no_tax <- phy %>%
  select(-taxa)

abun.phy.all <- melt(cbind(map.in,t(phy.no_tax)),id.vars = c("SampleID","Cave","Type","Elevation","Temperature","Humidity")) %>%
  group_by(variable) %>%
  mutate(mean = mean(value)*100) %>%
  select("variable","mean") %>%
  distinct(variable,.keep_all = T) %>%
  cbind(taxa = phy$taxa) %>%
  separate(taxa,c("kindom","phylum"),sep = ";") %>%
  separate(kindom,c("k","kindom"),sep = "__") %>%
  separate(phylum,c("p","phylum"),sep = "__") %>%
  select(-kindom,-k,-p,-variable) %>%
  mutate(phy = ifelse(mean <= 1,"Other",`phylum`)) %>%
  group_by(phy) %>%
  mutate(ra = sum(mean)) %>%
  distinct(phy,.keep_all = T) %>%
  select(-variable,-phylum,-mean)

abun.phy.all$phy <- factor(abun.phy.all$phy,levels = c("Proteobacteria","Actinobacteria","Firmicutes","Crenarchaeota","Gemmatimonadetes","Chloroflexi","Bacteroidetes","Other"))

phy.pie <- abun.phy.all %>%
  ggplot(aes("",ra,fill = phy)) + geom_bar(color = "white",stat = "identity",width = 1,alpha = 0.8) + coord_polar(theta = "y") + labs(x = "",y = "",title = "") + theme_minimal() + theme(axis.ticks = element_blank(),legend.title = element_blank(),axis.text.x = element_blank(), legend.position = "bottom") + scale_fill_brewer(palette = "Dark2")
phy.pie  

ggsave(phy.pie,filename = "phy_pie.pdf",height = 4,width = 4)

abun.phy <- melt(cbind(map.in,t(phy.no_tax)),id.vars = c("SampleID","Cave","Type","Elevation","Temperature","Humidity")) %>%
  group_by(Cave,variable) %>%
  mutate(mean = mean(value)*100) %>%
  group_by(variable) %>%
  mutate(core = ifelse(mean(mean) >= 1,"yes","no")) %>%
  select("Cave","variable","mean","core") %>%
  distinct(Cave,.keep_all = T) %>%
  ungroup() %>%
  spread(Cave,mean) %>%
  cbind(taxa = phy$taxa) %>%
  filter(core == "yes") %>%
  separate(taxa,c("kindom","phylum"),sep = ";") %>%
  separate(kindom,c("k","kindom"),sep = "__") %>%
  separate(phylum,c("p","phylum"),sep = "__") %>%
  select(-kindom,-k,-p,-core,-variable) %>%
  gather(cave,mean,-phylum)

abun.phy$phylum <- factor(abun.phy$phylum,levels = c("Proteobacteria","Actinobacteria","Firmicutes","Crenarchaeota","Gemmatimonadetes","Chloroflexi","Bacteroidetes"))

phy.p <- abun.phy %>%
  ggplot(aes(cave,phylum)) + theme_classic() + geom_tile(aes(fill = mean),color = "white") + geom_text(aes(label = round(mean,2)),color = ifelse(abun.phy$mean >= 1.1,"black","white")) + theme(axis.text.x = element_text(angle = 45,hjust = 1,vjust = 1),axis.title = element_blank()) + scale_fill_gradientn(colors = c("royalblue4","yellow","palevioletred1","indianred1"),values = c(0,0.02,0.3,1)) + theme(legend.position = "none")
phy.p

ggsave(phy.p,filename = "heatmap_phy.pdf",height = 4,width = 6)

fam <- read.table("in.taxa_summary/cave.in_otu_table_filtered_even_sorted_L5.txt",header = T,row.names = 1,sep = "\t") %>%
  rownames_to_column("taxa")

fam.no_tax <- fam %>%
  select(-taxa)

abun.fam <- melt(cbind(map.in,t(fam.no_tax)),id.var = c("SampleID","Cave","Type","Elevation","Temperature","Humidity")) %>%
  group_by(Cave,variable) %>%
  mutate(mean = mean(value)*100) %>%
  group_by(variable) %>%
  mutate(core = ifelse(mean(mean) >= 1,"yes","no")) %>%
  select("Cave","variable","mean","core") %>%
  distinct(Cave,.keep_all = T) %>%
  ungroup() %>%
  spread(Cave,mean) %>%
  cbind(taxa = fam$taxa) %>%
  filter(core == "yes") %>%
  separate(taxa,c("kindom","phylum","class","order","family"),sep = ";") %>%
  separate(kindom,c("k","kindom"),sep = "__") %>%
  separate(phylum,c("p","phylum"),sep = "__") %>%
  separate(class,c("c","class"),sep = "__") %>%
  separate(order,c("o","order"),sep = "__") %>%
  separate(family,c("f","family"),sep = "__") %>%
  select(-variable,-k,-p,-c,-o,-f,-core,-kindom) %>%
  gather(cave,mean,-family,-order,-class,-phylum) %>%
  mutate(family = ifelse(family == "","Unclassified",`family`)) %>%
  unite(taxa,phylum,class,order,family,sep = "/",remove = T) %>%
  group_by(taxa) %>%
  mutate(all.mean = mean(mean))
  
fam.p <- abun.fam %>%
  ggplot(aes(cave,taxa)) + theme_classic() + geom_tile(aes(fill = mean),color = "white") + geom_text(aes(label = round(mean,2)),color = ifelse(abun.fam$mean >= 1.5,"black","white")) + theme(axis.text.x = element_text(angle = 45,hjust = 1,vjust = 1),axis.title = element_blank()) + scale_fill_gradientn(colors = c("royalblue4","yellow","palevioletred1","indianred1"),values = c(0,0.1,0.3,1)) + theme(legend.position = "none")
fam.p

ggsave(fam.p,filename = "heatmap_fam.pdf",width = 9,height = 5)
```

## assembly pattern

### alpha diversity

```{r}
alpha <- read.table("alpha.txt",header = T,sep = "\t") %>%
  group_by(niche) %>%
  mutate(mean.no = mean(number),sd.no = sd(number),mean.cov = mean(coverage),sd.cov = sd(coverage),mean.chao1 = mean(chao1),sd.chao1 = sd(chao1),mean.shannon = mean(shannon),sd.shannon = sd(shannon))

shapiro.test(alpha$number) #p-value = 0.00453
shapiro.test(alpha$coverage) #p-value = 3.753e-05
shapiro.test(alpha$chao1) #p-value = 0.0003642
shapiro.test(alpha$shannon) #p-value = 0.08756
bartlett.test(shannon ~ niche,alpha) #p-value = 0.02133

kruskal.test(number ~ niche,alpha)
#chi-squared = 8.062, df = 3, p-value = 0.04475
dunnTest(number ~ niche,alpha,method = "none")
# Comparison        P.unadj
# Air - Rock        0.84881995
# Air - Sediment    0.03422181
# Rock - Sediment   0.02092284
# Air - Water       0.10923249
# Rock - Water      0.07839066
# Sediment - Water  0.88301636

kruskal.test(coverage ~ niche,alpha)
#chi-squared = 12.864, df = 3, p-value = 0.004939
dunnTest(coverage ~ niche,alpha,method = "none")
# Comparison        P.unadj
# Air - Rock       0.582561668
# Air - Sediment   0.010248637
# Rock - Sediment  0.044017219
# Air - Water      0.003358856
# Rock - Water     0.013287832
# Sediment - Water 0.413849385

kruskal.test(chao1 ~ niche,alpha)
#chi-squared = 10.442, df = 3, p-value = 0.01516
dunnTest(chao1 ~ niche,alpha,method = "none")
# Comparison          P.unadj
# Air - Rock        0.67726804
# Air - Sediment    0.01124052
# Rock - Sediment   0.03433662
# Air - Water       0.01879989
# Rock - Water      0.04512034
# Sediment - Water  0.79660546

kruskal.test(shannon ~ niche,alpha)
#chi-squared = 17.205, df = 3, p-value = 0.0006414
dunnTest(shannon ~ niche,alpha,method = "none")
# Comparison          P.unadj
# Air - Rock       0.140440442
# Air - Sediment   0.000042773
# Rock - Sediment  0.009113914
# Air - Water      0.127261425
# Rock - Water     0.764435792
# Sediment - Water 0.062611080

bartlett.test(shannon ~ cave,alpha) #p-value = 0.1088

cave.aov <- aov(shannon ~ cave,alpha)
summary(cave.aov)
#             Df Sum Sq Mean Sq F value Pr(>F)
#cave          7  14.02   2.003   1.352  0.232
#Residuals   120 177.79   1.482               

kruskal.test(chao1 ~ cave,alpha)
#chi-squared = 8.475, df = 7, p-value = 0.2926

shannon.p <- alpha %>%
  ggplot(aes(cave,shannon,color = cave)) + stat_boxplot(geom = "errorbar",width = 0.2,aes(color = cave)) + geom_boxplot(alpha = 1,outlier.size = 0,size = 0.9, width = 0.5) + theme_bw() + labs(x = "Cave",y = "Shannon indices") + theme(axis.text.x = element_text(angle = 45,vjust = 1,hjust = 1),axis.title.x = element_blank()) + scale_color_brewer(palette = "Set2") + theme(legend.position = "none")
shannon.p

ggsave(shannon.p,filename = "shannon_cave.pdf",height = 4,width = 7)

chao1.p <- alpha %>%
  ggplot(aes(cave,chao1,color = cave)) + stat_boxplot(geom = "errorbar",width = 0.2,aes(color = cave)) + geom_boxplot(alpha = 1,outlier.size = 0,size = 0.9, width = 0.5) + theme_bw() + labs(x = "Cave",y = "Chao1 indices") + theme(axis.text.x = element_text(angle = 45,vjust = 1,hjust = 1),axis.title.x = element_blank()) + scale_color_brewer(palette = "Set2") + theme(legend.position = "none")
chao1.p

ggsave(chao1.p,filename = "chao1_cave.pdf",height = 4,width = 7)
```

### beta diverssity

```{r}
library(vegan)

pcoa <- function(x,m) {
  pco <- cmdscale(x, k = 3, eig = T)
  pco.axes <- as.data.frame(pco$points)
  colnames(pco.axes) <- c("x", "y", "z")
  pco.axes <- cbind(pco.axes,m[match(row.names(pco.axes),m$SampleID),])
  pco.eig <- eigenvals(pco) / sum(eigenvals(pco))
  pco.res <- list(axes = pco.axes, eig = pco.eig)
  return(pco.res)
}

weighted <- read.table("in.beta/weighted_unifrac_cave.in_otu_table_css.txt", sep="\t", header=T, check.names=F)

pcoa.weighted <- pcoa(weighted,map.in)
head(pcoa.weighted$eig)
#0.38130601 0.15856969 0.08265793 0.05239138 0.03850398 0.03476222

type.pcoa <- pcoa.weighted$axes %>%
  ggplot(aes(x,y,color = Type)) + geom_point(size = 2.5,alpha  = 0.8) + theme_bw() +  theme(panel.grid = element_blank(),legend.position = "bottom",legend.title = element_blank()) + scale_color_manual(values = c("#c72e29","#016392","#fb832d","#098154")) + stat_ellipse(level = 0.95,linetype = 5) + labs(x = "PCo1 (38.13%)",y = "PCo2 (15.86%)") + geom_hline(aes(yintercept = 0)) + geom_vline(aes(xintercept = 0))

type.pcoa
ggsave(type.pcoa,filename = "pcoa_2.pdf",width = 4,height = 4)
```

## leaner regression

```{r}
dis <- as.matrix(vegdist(log2(t(counts.in_css[,1:128])+1)))

dis[upper.tri(dis,diag = T)] <- NA

geo_xy_eu[upper.tri(geo_xy_eu,diag = T)] <- NA

geo_xy_eu_m <- melt(geo_xy_eu) %>%
  na.omit() %>%
  mutate(dist = value)

dis_m <- melt(dis) %>%
  na.omit() %>%
  inner_join(map.in,by = c("Var1" = "SampleID")) %>%
  inner_join(map.in, by = c("Var2" = "SampleID")) %>%
  cbind(dist = geo_xy_eu_m$dist)

lm.G <- dis_m %>%
  ggplot(aes(dist,1-value)) + geom_point(alpha = 0.5) + theme_bw() + stat_smooth(method = "lm",se = F) + labs(x = "Geographical distance",y = "1-Bray dissimilarity")
lm.G
# Multiple R-squared:  0.01071,	Adjusted R-squared:  0.01059
# F-statistic: 88.01 on 1 and 8126 DF,  p-value: < 2.2e-16
ggsave(lm.G,filename = "lm_G.pdf",width = 4,height = 3)

lm.T <- dis_m %>%
  filter(Temperature.x == Temperature.y) %>%
  ggplot(aes(Temperature.x,1-value)) + geom_point(alpha = 0.5) + theme_bw() + stat_smooth(method = "lm",se = F) + labs(x = "Temperature",y = "1-Bray dissimilarity")
lm.T
# y = 0.0015171x + 0.142156
# Multiple R-squared:  0.008247,	Adjusted R-squared:  0.007152
# F-statistic: 7.534 on 1 and 906 DF,  p-value: 0.006174
ggsave(lm.T,filename = "lm_T.pdf",width = 4,height = 3)

lm.H <- dis_m %>%
  filter(Humidity.x == Humidity.y) %>%
  ggplot(aes(Humidity.x,1-value)) + geom_point(position = position_jitterdodge(),alpha = 0.5) + theme_bw() + stat_smooth(method = "lm",se = F) + labs(x = "Air humidity",y = "1-Bray dissimilarity")
lm.H
# y = -0.16205x + 0.30599
# Multiple R-squared:  0.005069,	Adjusted R-squared:  0.004242
# F-statistic: 6.125 on 1 and 1202 DF,  p-value: 0.01347
ggsave(lm.H,filename = "lm_H.pdf",width = 4,height = 3)
```

## variation

```{r}
phyla <- read.table("phyla.txt",header = T,row.names = 1,sep = "\t") %>%
  gather("phylum","RA",-niche) %>%
  group_by(niche,phylum) %>%
  mutate(mean = mean(RA)) %>%
  distinct(niche,phylum,.keep_all = T) %>%
  select(-RA)

phyla$phylum <- factor(phyla$phylum,levels = c("Alphaproteobacteria","Betaproteobacteria","Gammaproteobacteria","Deltaproteobacteria","Actinobacteria","Firmicutes","Crenarchaeota","Gemmatimonadetes","Chloroflexi","Bacteroidetes","Other"))

phyla.p <- phyla %>%
  ggplot(aes(niche,mean,fill = phylum)) + geom_bar(stat = "identity",position = "fill") + theme_minimal() + scale_fill_brewer(palette = "Set3") + labs(x = "",y = "Relative Abundance") + theme(legend.title = element_blank(),legend.position = "bottom") + coord_flip()
phyla.p

ggsave(phyla.p,filename = "phyla_stacking.pdf",height = 3,width = 5)
```

```{r}
library(VennDiagram)

counts_clean.in_no_tax <- counts_clean.in %>%
  select(-taxonomy)

taxa <- counts_clean.in %>%
  rownames_to_column("OTU") %>%
  select(OTU,taxonomy)

counts_clean.in_map <- melt(cbind(map.in,t(counts_clean.in_no_tax)),id.vars = c("SampleID","Cave","Type","Elevation","Temperature","Humidity")) %>%
  left_join(taxa,by = c("variable" = "OTU"))

air <- counts_clean.in_map %>%
  filter(Type == "Air") %>%
  select(variable,SampleID,value,taxonomy) %>%
  group_by(SampleID,variable) %>%
  mutate(keep = ifelse(sum(value) == 0,"no","yes")) %>%
  filter(keep == "yes") %>%
  select(-keep) %>%
  spread(SampleID,value,fill = 0)

rock <- counts_clean.in_map %>%
  filter(Type == "Rock") %>%
  select(variable,SampleID,value,taxonomy) %>%
  group_by(SampleID,variable) %>%
  mutate(keep = ifelse(sum(value) == 0,"no","yes")) %>%
  filter(keep == "yes") %>%
  select(-keep) %>%
  spread(SampleID,value,fill = 0)

sediment <- counts_clean.in_map %>%
  filter(Type == "Sediment") %>%
  select(variable,SampleID,value,taxonomy) %>%
  group_by(SampleID,variable) %>%
  mutate(keep = ifelse(sum(value) == 0,"no","yes")) %>%
  filter(keep == "yes") %>%
  select(-keep) %>%
  spread(SampleID,value,fill = 0)

water <- counts_clean.in_map %>%
  filter(Type == "Water") %>%
  select(variable,SampleID,value,taxonomy) %>%
  group_by(SampleID,variable) %>%
  mutate(keep = ifelse(sum(value) == 0,"no","yes")) %>%
  filter(keep == "yes") %>%
  select(-keep) %>%
  spread(SampleID,value,fill = 0)

venn <- venn.diagram(list(Air = air$variable,Rock = rock$variable,Sediment = sediment$variable,Water = water$variable),filename = NULL,alpha = 0.7,lty = 1,col = c("#c72e29","#016392","#fb832d","#098154"))
grid.draw(venn)
```

### indicspecies

```{r}
library(indicspecies)

gen <- read.table("in.taxa_summary/cave.in_otu_table_filtered_even_sorted_L6.txt",header = T,row.names = 1,sep = "\t") %>%
  rownames_to_column("taxa")

gen.no_tax <- gen %>%
  select(-taxa)

ind.gen <- melt(cbind(map.in,t(gen.no_tax)),id.var = c("SampleID","Cave","Type","Elevation","Temperature","Humidity")) %>%
  group_by(Type,variable) %>%
  mutate(med = median(value)*100) %>%
  group_by(variable) %>%
  mutate(core = ifelse(max(med) >= 0.5,"yes","no")) %>%
  select("SampleID","variable","core","value") %>%
  ungroup() %>%
  spread(SampleID,value) %>%
  cbind(taxa = gen$taxa) %>%
  filter(core == "yes") %>%
  separate(taxa,c("kindom","phylum","class","order","family","genus"),sep = ";") %>%
  separate(kindom,c("k","kindom"),sep = "__") %>%
  separate(phylum,c("p","phylum"),sep = "__") %>%
  separate(class,c("c","class"),sep = "__") %>%
  separate(order,c("o","order"),sep = "__") %>%
  separate(family,c("f","family"),sep = "__") %>%
  separate(genus,c("g","genus"),sep = "__") %>%
  select(-variable,-k,-p,-c,-o,-f,-g,-core,-kindom) %>%
  mutate(keep = ifelse(family == "","no","yes")) %>%
  filter(keep == "yes") %>%
  mutate(genus = ifelse(genus == "","Unclassified",`genus`)) %>%
  unite(taxa,phylum,class,order,family,genus,sep = "/",remove = T) %>%
  select(-keep)

ind.gen.t <- ind.gen %>%
  column_to_rownames("taxa") %>%
  t(.) %>%
  as.data.frame(.)
  
ind.gro <- c(rep(1,6),rep(2,6),rep(3,6),rep(4,3),rep(1,4),rep(2,5),rep(3,6),rep(4,3),rep(1,5),rep(2,5),rep(3,5),rep(4,1),rep(1,4),rep(2,4),rep(3,4),rep(4,2),rep(1,5),rep(2,5),rep(3,4),rep(4,3),rep(1,4),rep(2,4),rep(3,4),rep(4,2),rep(1,5),rep(2,5),rep(3,5),rep(4,3),rep(1,3,),rep(2,2),rep(3,3),rep(4,2))

ind <- multipatt(ind.gen.t,ind.gro,func = "IndVal", duleg=TRUE, control = how(nperm=9999))
summary(ind,indvalcomp = TRUE)

holm.pro <- p.adjust(ind$sign$p.value,"holm")
holm.pro

ind.p.holm <- cbind(ind$sign,holm.pro)
```

## network

```{r}
gen <- read.table("../190110_compile/in.taxa_summary/cave.in_otu_table_filtered_even_sorted_L6.txt",sep = "\t",header = T,row.names = 1)

gen_net <- gen[rowSums(gen) >= 0.05,]
gen_net <- gen_net[2:nrow(gen_net),]
gen_sum <- gen_net %>%
  mutate(sum = rowSums(gen_net))
row.names(gen_sum) <- row.names(gen_net)

gen_sum %>%
  write.table("gen_sum.txt",sep = "\t",row.names = T)

air <- cbind(gen_net[,1:6],gen_net[,22:25],gen_net[,40:44],gen_net[,56:59],gen_net[,70:74],gen_net[,87:90],gen_net[,101:105],gen_net[,119:121])
air <- air %>%
  mutate(sum = rowSums(air),taxa = row.names(gen_net)) %>%
  filter(sum != 0)
air.t <- t(air[,1:36])

air %>%
  write.table("air_sum.txt",sep = "\t",row.names = T)

rock <- cbind(gen_net[,7:12],gen_net[,26:30],gen_net[,45:49],gen_net[,60:63],gen_net[,75:79],gen_net[,91:94],gen_net[,106:110],gen_net[,122:123])
rock <- rock %>%
  mutate(sum = rowSums(rock),taxa = row.names(gen_net)) %>%
  filter(sum != 0)
rock.t <- t(rock[,1:36])

rock %>%
  write.table("rock_sum.txt",sep = "\t",row.names = T)

sed <- cbind(gen_net[,13:18],gen_net[,31:36],gen_net[,50:54],gen_net[,64:67],gen_net[,80:83],gen_net[,95:98],gen_net[,111:115],gen_net[,124:126])
sed <- sed %>%
  mutate(sum = rowSums(sed),taxa = row.names(gen_net)) %>%
  filter(sum != 0)
sed.t <- t(sed[,1:37])

sed %>%
  write.table("sediment_sum.txt",sep = "\t",row.names = T)

water <- cbind(gen_net[,19:21],gen_net[,37:39],gen_net[,55],gen_net[,68:69],gen_net[,84:86],gen_net[,99:100],gen_net[,116:118],gen_net[,127:128])
water <- water %>%
  mutate(sum = rowSums(water),taxa = row.names(gen_net)) %>%
  filter(sum != 0)
water.t <- t(water[,1:19])

water %>%
  write.table("water_sum.txt",sep = "\t",row.names = T)
```

```{r}
library(Hmisc)
library(igraph)
library(fdrtool)

air.corr <- rcorr(air.t,type = "spearman")
air.r <- air.corr$r
air.p <- air.corr$P
diag(air.p) <- 1
air.pp <- as.vector(air.p)
air.qval <- fdrtool(air.pp,statistic = "pvalue")
air.q <- air.qval$qval
air.q <- matrix(air.q,ncol(air.p),ncol(air.p))
air.q[air.q > 0.01] <- 0
air.q[air.q <= 0.01 & air.q > 0] <- 1
air.r[air.r < 0.6] <- 0
air.g <- air.r * air.q
air.g <- graph.adjacency(air.g,weighted = T,mode = "undirected")
air.g <- simplify(air.g)
air.g <- delete.vertices(air.g,names(degree(air.g)[degree(air.g) == 0]))
air.d <- degree(air.g)
write.graph(air.g,"air_gen_network.gml",format = "gml")

rock.corr <- rcorr(rock.t,type = "spearman")
rock.r <- rock.corr$r
rock.p <- rock.corr$P
diag(rock.p) <- 1
rock.pp <- as.vector(rock.p)
rock.qval <- fdrtool(rock.pp,statistic = "pvalue")
rock.q <- rock.qval$qval
rock.q <- matrix(rock.q,ncol(rock.p),ncol(rock.p))
rock.q[rock.q > 0.01] <- 0
rock.q[rock.q <= 0.01 & rock.q > 0] <- 1
rock.r[rock.r < 0.6] <- 0
rock.g <- rock.r * rock.q
rock.g <- graph.adjacency(rock.g,weighted = T,mode = "undirected")
rock.g <- simplify(rock.g)
rock.g <- delete.vertices(rock.g,names(degree(rock.g)[degree(rock.g) == 0]))
rock.d <- degree(rock.g)
write.graph(rock.g,"rock_network.gml",format = "gml")

sed.corr <- rcorr(sed.t,type = "spearman")
sed.r <- sed.corr$r
sed.p <- sed.corr$P
diag(sed.p) <- 1
sed.pp <- as.vector(sed.p)
sed.qval <- fdrtool(sed.pp,statistic = "pvalue")
sed.q <- sed.qval$qval
sed.q <- matrix(sed.q,ncol(sed.p),ncol(sed.p))
sed.q[sed.q > 0.01] <- 0
sed.q[sed.q <= 0.01 & sed.q > 0] <- 1
sed.r[sed.r < 0.6] <- 0
sed.g <- sed.r * sed.q
sed.g <- graph.adjacency(sed.g,weighted = T,mode = "undirected")
sed.g <- simplify(sed.g)
sed.g <- delete.vertices(sed.g,names(degree(sed.g)[degree(sed.g) == 0]))
sed.d <- degree(sed.g)
write.graph(sed.g,"sed_network.gml",format = "gml")

water.corr <- rcorr(water.t,type = "spearman")
water.r <- water.corr$r
water.p <- water.corr$P
diag(water.p) <- 1
water.pp <- as.vector(water.p)
water.qval <- fdrtool(water.pp,statistic = "pvalue")
water.q <- water.qval$qval
water.q <- matrix(water.q,ncol(water.p),ncol(water.p))
water.q[water.q > 0.01] <- 0
water.q[water.q <= 0.01 & water.q > 0] <- 1
water.r[water.r < 0.6] <- 0
water.g <- water.r * water.q
water.g <- graph.adjacency(water.g,weighted = T,mode = "undirected")
water.g <- simplify(water.g)
water.g <- delete.vertices(water.g,names(degree(water.g)[degree(water.g) == 0]))
water.d <- degree(water.g)
write.graph(water.g,"water_network.gml",format = "gml")

corr <- rcorr(t(gen_net),type = "spearman")
r <- corr$r
p <- corr$P
diag(p) <- 1
pp <- as.vector(p)
qval <- fdrtool(pp,statistic = "pvalue")
q <- qval$qval
q <- matrix(q,ncol(p),ncol(p))
q[q > 0.01] <- 0
q[q <= 0.01 & q > 0] <- 1
r[r < 0.6] <- 0
g <- r * q
g <- graph.adjacency(g,weighted = T,mode = "undirected")
g <- simplify(g)
g <- delete.vertices(g,names(degree(g)[degree(g) == 0]))
write.graph(g,"gen_network.gml",format = "gml")
```

```{r}
air.node <- read.csv("air_network_node.csv",sep = ",",header = T) %>%
  left_join(air,by = c("name" = "taxa")) %>%
  select(Id,Label,timeset,name,sum)

write.csv(air.node,file = "air_node_anno.csv")

rock.node <- read.csv("rock_network_node.csv",sep = ",",header = T) %>%
  left_join(rock,by = c("name" = "taxa")) %>%
  select(Id,Label,timeset,name,sum)

write.csv(rock.node,file = "rock_node_anno.csv")

sed.node <- read.csv("sediment_network_node.csv",sep = ",",header = T) %>%
  left_join(rock,by = c("name" = "taxa")) %>%
  select(Id,Label,timeset,name,sum)

write.csv(sed.node,file = "sediment_node_anno.csv")

water.node <- read.csv("water_network_node.csv",sep = ",",header = T) %>%
  left_join(water,by = c("name" = "taxa")) %>%
  select(Id,Label,timeset,name,sum)

write.csv(water.node,file = "water_node_anno.csv")
```

